Support hermetic index store data via path remappings

DavidGoldman opened the pull request:
- When indexing, `-fdebug-prefix-map` will be used to remap
  paths so the computed index data will be hermetic. This
  includes the unit files' names as well as all paths embedded
  in the unit files.

- When reading, a reverse mapping can be used e.g. via
  `indexstore_store_create_with_prefix_mapping` to automatically
  remap the hermetic index data to the correct local paths.

DavidGoldman committed:
Support remapping index store paths

- When indexing, `-fdebug-prefix-map` will be used to remap
  paths so the computed index data will be hermetic. This
  includes the unit files' names as well as all paths embedded
  in the unit files.

- When reading, a reverse mapping can be used e.g. via
  `indexstore_store_create_with_prefix_mapping` to automatically
  remap the hermetic index data to the correct local paths.

DavidGoldman commented:
@swift-ci Please test

keith commented:
@swift-ci please test

DavidGoldman commented:
Updated the tests to confirm that the unit file names are deterministic by including the hash in the expected output

bnbarham commented on clang/include/clang/Index/IndexDataStore.h:
```diff
@@ -26,9 +27,14 @@ class IndexDataStore {
   static std::unique_ptr<IndexDataStore>
-    create(StringRef IndexStorePath, std::string &Error);
+    create(StringRef IndexStorePath,
+      std::map<std::string, std::string, std::greater<std::string>>
+          PrefixMap,
```
I think it'd be nice to extract this out into a separate data structure. IIUC there's no need for this to be a map other than that it's convenient (though not really) to pass around. So if we had a separate class for this it could internally just store a SmallVector of `<struct with original and remap path>`.

akyrtzi replied:
I agree, and as an input to the API I think conceptually it should be an array of string pairs. Whether then the underlying implementation uses a `std::map` or a `llvm::StringMap` for the lookups, it would be an implementation detail.

bnbarham replied:
FWIW I don't think a "lookup" is ever actually performed, we only ever iterate over them all and check whether there's a prefix match (hence my suggestion of a SmallVector internally).

DavidGoldman replied:
I was mirroring how Clang stores its prefix maps: https://cs.github.com/llvm/llvm-project/blob/0f304ef0170231b860a249f34e07f50686392253/clang/lib/CodeGen/CGDebugInfo.h#L87 and https://cs.github.com/llvm/llvm-project/blob/0f304ef0170231b860a249f34e07f50686392253/clang/include/clang/Basic/LangOptions.h#L421. I can change this to be more like how Swift does it (they're different, like you mentioned)

bnbarham replied:
Ideally those wouldn't be maps either 😅. So these are checked in alphabetical order rather than the order they were given on the CLI?

DavidGoldman replied:
Yes, that's my understanding. I think it would make sense to clean up, but not sure it's in scope for here. Thoughts?

bnbarham replied:
I'd say it's worth making whatever you add usable in those cases, but we'd want to fix that upstream anyway (so no, not in scope).

bnbarham commented on clang/include/indexstore/indexstore.h:
```diff
@@ -133,6 +133,11 @@ typedef void *indexstore_t;
+INDEXSTORE_PUBLIC indexstore_t
+indexstore_store_create_with_prefix_mapping(
+    const char *store_path, const char **PrefixMappings,
+    size_t NumMappings,indexstore_error_t *error);
```
Nit: run clang-format.

akyrtzi replied:
To be more flexible about evolving the C API I'd recommend to avoid adding an ad-hoc `indexstore_store_create_with_prefix_mapping` (ideally we'd avoid the situation of adding new variants of that function whenever we find a need to pass something additional).

I'd suggest to introduce a new opaque object, like `indexstore_creation_options_t`, with C functions to set values for it, then introduce a `indexstore_store_create_with_options` function.

akyrtzi replied:
You should also increment the `INDEXSTORE_VERSION_MINOR` value in the header.

DavidGoldman replied:
SGTM, I'll go ahead and do that, that makes sense for future proofing.

DavidGoldman replied:
Done

bnbarham commented on clang/lib/Index/IndexUnitWriter.cpp:
```diff
@@ -89,8 +104,16 @@ class IndexUnitWriter::PathStorage {
-      dirPath.Dir.Offset = getPathOffset(dirStr);
-      dirPath.Dir.Size = dirStr.size();
+
+      if (dirPath.PrefixKind != UNIT_PATH_PREFIX_NONE) {
```
Would it make sense to add a new prefix kind? Ie. `UNIT_PATH_PREFIX_MAPPED`? That would also allow us to eg. output a warning if a mapping isn't provided when reading, though perhaps we actually *don't* want that. Ie. the path being mapped to could be valid as well.

DavidGoldman replied:
Yeah, we could, at the downside of needing a version bump, plus some slight duplication since most things are relative to the workdir/sysroot in practice.

bnbarham replied:
I think I convinced myself it's wrong anyway. I don't think the reader should know whether it was mapped or not in the prefix map case.

bnbarham commented on clang/tools/IndexStore/IndexStore.cpp:
```diff
@@ -98,14 +100,34 @@ indexstore_format_version(void) {
+}
+
+indexstore_t
+indexstore_store_create_with_prefix_mapping(const char *store_path, const char **PrefixMappings,
```
Could add a struct for a mapping and then pass in `const indexstore_path_mapping_t *` rather than the concatenation/splitting with `=`.

DavidGoldman replied:
Went with the `indexstore_store_create_with_options` approach

DavidGoldman committed:
Swap to indexstore_store_create_with_options

Also increment INDEXSTORE_VERSION_MINOR

akyrtzi commented on clang/include/indexstore/indexstore.h:
```diff
@@ -129,10 +129,24 @@ INDEXSTORE_PUBLIC unsigned
 indexstore_store_create(const char *store_path, indexstore_error_t *error);
 
+INDEXSTORE_PUBLIC indexstore_t
+indexstore_store_create_with_options(const char *store_path, indexstore_creation_options_t options,
```
Could you add a doc-comment to clarify that `options` can be `NULL`?

DavidGoldman replied:
Done

DavidGoldman committed:
Add documentation comments to indexstore.h

Also avoid storing relative paths in the index data - make them
absolute first, which means they can be stored using
UNIT_PATH_PREFIX_WORKDIR.

DavidGoldman commented:
Also went ahead and made sure paths are absolutized in IndexUnitWriter --> meaning they will be stored relative to UNIT_PATH_PREFIX_WORKDIR.

DavidGoldman committed:
indexstore_creation_options_create() --> indexstore_creation_options_create(void)

DavidGoldman commented:
@swift-ci please test

DavidGoldman committed:
Don't make stdout absolute

DavidGoldman committed:
Add PathRemapper utility class

- The PathRemapper preserves insertion order of the remappings instead
  of sorting

DavidGoldman commented:
Closing, will re-open with cherry-picks from stable.
